rfcs/2022--01-06-8596-tcp-source-backpressure.md

```diff
@@ -0,0 +1,160 @@
+
+Instead of trying to control the number of in-flight requests at the connection level, the requests can just be controlled directly.
+The number of TCP connections could stay unlimited, but there is a check before each request is processed that
+can limit how many are actually in-flight. The goal here is to pick a limit to the number of in-flight requests and
+only allow new ones to be processed if you are below that. This can't be limited to an _exact_ number of events, because you
+don't know how many events are in a specific request until it has been decoded. But you can use a limit in the form of
+`x events + y requests`. This would attempt to limit the in-flight events to `x`, plus `y` full requests of events.
+
+The main drawback to this approach is that you have to be able to accept a request before you know how many events
+are in that request, and the total in-flight count can't be updated until after the decoding has finished. A concurrency
+limit must be placed on request decoding (equal to `y` above) in order to limit the overall requests in-flight.
+
+This leads us to the question of what should the value of `x` and `y` be? `x` (the number of events in-flight) can probably
+be set to 0 and ignored, since we already have event buffering in other parts of the system (the pipeline and sink buffers).
+Choosing `y` is balancing performance with memory usage.
+
+I propose that `y` is a value dynamically determined by the request size of previous requests.
+The "request size" will simply be defined as the number of events contained in a request. The running-average of
+these will be used to estimate the size of future requests. Now a target number of in-flights requests can be chosen.
+`y` will be set to a value such that actual in-flight messages + estimated in-flight messages is less than the target.
+Once a request is decoded the actual in-flight messages can be increased, and as they are flushed into the pipeline
+the count will be decreased. The target value will be a constant value somewhat arbitrarily chosen based on perf tests.
+The exact value will be proposed in the final implementation.
+
+A dynamic value for `y` is useful because the size of the requests can vary drastically depending on the source / protocol used.
+Each request may only contain a single event, or as noted above can contain tens of thousands. Using a constant
+value would likely penalize sources using small requests by artificially limiting concurrency, or allowing
+sources with large requests to have too many events in memory.
+
+On top of the dynamic value for `y`, some hard limits will be added. Regardless of the current request size estimate,
+1 request must always be allowed to be processed when the actual in-flight message count is zero to ensure forward
+progress can be made. In addition, an upper-bound will be set equal to the number of CPU's on the system. A
+value higher than this has no benefit, but would use more memory.
+
```

Comment:
To double check my understanding, is this effectively how the dynamic limit works?
- 1 <= `y` <= # CPUs
- After a new request is decoded, adjust `y` if needed such that `y` * `{moving avg # of events/request}` < `{pre-determined target # of in-flight events}` - `{current # of in-flight events}`

And `y` starts at `1`?

Reply:
Yes, that is correct